diesel/src/query_builder/combination_clause.rs

```diff
@@ -0,0 +1,206 @@
+//! Combine queries using a combinator like `UNION`, `INTERSECT` or `EXPECT`
+//! with or without `ALL` rule for duplicates
+
+use crate::backend::Backend;
+use crate::query_builder::insert_statement::InsertFromSelect;
+use crate::query_builder::{AstPass, Query, QueryFragment, QueryId};
+use crate::QueryDsl;
+use crate::{Insertable, QueryResult};
+use crate::{RunQueryDsl, Table};
+use std::marker::Sized;
+
+#[derive(Debug, Copy, Clone, QueryId)]
+#[must_use = "Queries are only executed when calling `load`, `get_result` or similar."]
+/// Combine queries using a combinator like `UNION`, `INTERSECT` or `EXPECT`
+/// with or without `ALL` rule for duplicates
+pub struct Combination<Combinator, Rule, Source, Rhs> {
+    combinator: Combinator,
+    duplicate_rule: Rule,
+    source: Source,
+    rhs: Rhs,
+}
+
+impl<Combinator, Rule, Source, Rhs> Combination<Combinator, Rule, Source, Rhs> {
+    /// Create a new combination
+    pub fn new(combinator: Combinator, duplicate_rule: Rule, source: Source, rhs: Rhs) -> Self {
+        Combination {
+            combinator,
+            duplicate_rule,
+            source,
+            rhs,
+        }
+    }
+}
+
+impl<Combinator, Rule, Source, Rhs> Query for Combination<Combinator, Rule, Source, Rhs>
+where
+    Source: Query,
+    Rhs: Query<SqlType = Source::SqlType>,
+{
+    type SqlType = Source::SqlType;
+}
+
+impl<Combinator, Rule, Source, Rhs, DB> QueryFragment<DB>
+    for Combination<Combinator, Rule, Source, Rhs>
+where
+    Combinator: QueryFragment<DB>,
+    Rule: QueryFragment<DB>,
+    Source: QueryFragment<DB>,
+    Rhs: QueryFragment<DB>,
+    DB: Backend,
+{
+    fn walk_ast(&self, mut out: AstPass<DB>) -> QueryResult<()> {
+        self.source.walk_ast(out.reborrow())?;
+        self.combinator.walk_ast(out.reborrow())?;
+        self.duplicate_rule.walk_ast(out.reborrow())?;
+        self.rhs.walk_ast(out.reborrow())?;
+        Ok(())
+    }
+}
+
+impl<Combinator, Rule, Source, Rhs, Conn> RunQueryDsl<Conn>
+    for Combination<Combinator, Rule, Source, Rhs>
+{
+}
+
+impl<Combinator, Rule, Source, Rhs> QueryDsl for Combination<Combinator, Rule, Source, Rhs> {}
```

Comment:
I don't think that is currently useful as it does not provide any actual method from `QueryDsl` yet. For that we would need to implement some of the traits from `query_dsl::methods`

Reply:
I hit a limitation when trying to implement an `ORDER BY` for combination queries, as explained in [Pg docs](https://www.postgresql.org/docs/13/sql-select.html#SQL-ORDERBY) : 
> A limitation of this feature is that an ORDER BY clause applying to the result of a UNION, INTERSECT, or EXCEPT clause can only specify an output column name or number, not an expression.

That means the following query is invalid (`ERROR:  42P01: missing FROM-clause entry for table "users"`) : 
```sql
SELECT "users"."name"
FROM "users"
UNION
SELECT "animals"."name"
FROM "animals"
WHERE "animals"."name" IS NOT NULL
ORDER BY "users"."name"; -- "name" alone or 1 would work
```
Using the column name only doesn't look like a good idea because that can lead to conflicts, so that leave us with the numerical ordering. Any clues on how to get the ordering of a column in a select clause ?

Reply:
What about just not implement `OrderDsl` for the initial implementation? I mean the first version supporting this in diesel must not be complete, but can have some unpolished cases.
That written, I think correctly implementing that would require something like complete aliasing support for column names, in such a way that this columns can also be used in order/where clauses. At least the same infrastructure enabling something like that would be useful here. I would assume that this is something that will not implemented in the next few version, so I would just go with an initial version that does not support ordering for now.

Reply:
Ordering would be really handy in tests, what do you think of an undocumented `.positional_order_by(1.asc())` ?
If I can make if work, of course :wink:

Reply:
That would be a good solution, as long as it is clear that this method is not part of the public API. Maybe just something like done with `group_by` on the 1.x branch here: https://github.com/diesel-rs/diesel/blob/bec389a1ea329c161fc7a8fb8df8bea29a3432bf/diesel/src/query_dsl/group_by_dsl.rs
(So `#[doc(hidden)]` + an explicit comment in the source code that this is not part of the public API.)